File: go/internal/agent/setup/session_firestore.go
package setup

import (
	"context"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"iter"
	"strings"
	"time"

	"cloud.google.com/go/firestore"
	"github.com/google/uuid"
	"google.golang.org/adk/session"
	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/status"
)

// firestoreSessionService is a custom session.Service backed by Firestore — the cloud
// durable session store (adk-go ships only inmemory/database/vertexai). It mirrors the
// in-memory service's semantics: app:/user:/temp: state scopes, partial-event skipping,
// and event filtering on read. State scope routing replicates internal/sessionutils
// (which is not importable). Events are stored as JSON because their genai content does
// not map cleanly onto Firestore's value types.
type firestoreSessionService struct {
	client          *firestore.Client
	sessions, appSt string
	userSt          string
}

// NewFirestoreSessionService opens a Firestore-backed session service. project may be ""
// to detect it from ADC / GOOGLE_CLOUD_PROJECT. Close releases the client.
func NewFirestoreSessionService(ctx context.Context, project, prefix string) (*firestoreSessionService, error) {
	if project == "" {
		project = firestore.DetectProjectID
	}
	client, err := firestore.NewClient(ctx, project)
	if err != nil {
		return nil, fmt.Errorf("firestore client: %w", err)
	}
	return &firestoreSessionService{
		client:   client,
		sessions: prefix + "_sessions",
		appSt:    prefix + "_app_state",
		userSt:   prefix + "_user_state",
	}, nil
}

// Close releases the underlying Firestore client.
func (s *firestoreSessionService) Close() error { return s.client.Close() }

var _ session.Service = (*firestoreSessionService)(nil)

// --- Firestore document shapes ---

// sessionDoc is the persisted session: session-scoped state (app/user state live in their
// own collections). Events live in an "events" sub-collection (see eventDoc) rather than an
// array field, so a long-lived session cannot blow Firestore's 1 MiB per-document limit.
type sessionDoc struct {
	AppName   string         `firestore:"app_name"`
	UserID    string         `firestore:"user_id"`
	SessionID string         `firestore:"session_id"`
	State     map[string]any `firestore:"state"`
	NextSeq   int64          `firestore:"next_seq"` // next event sequence number
	UpdatedAt time.Time      `firestore:"updated_at"`
}

// eventDoc is one event in a session's "events" sub-collection, ordered by Seq.
type eventDoc struct {
	Seq       int64     `firestore:"seq"`
	Timestamp time.Time `firestore:"timestamp"`
	Blob      string    `firestore:"blob"` // JSON-encoded session.Event
}

type stateDoc struct {
	State map[string]any `firestore:"state"`
}

// --- key encoding ---

// encodeKey builds a Firestore-safe document id from its parts (base64url of a
// delimiter-joined key), so arbitrary app/user/session ids cannot collide or contain
// illegal characters.
func encodeKey(parts ...string) string {
	return base64.RawURLEncoding.EncodeToString([]byte(strings.Join(parts, "\x1f")))
}

// --- state scope helpers (replicate internal/sessionutils) ---

func extractStateDeltas(delta map[string]any) (app, user, sess map[string]any) {
	app, user, sess = map[string]any{}, map[string]any{}, map[string]any{}
	for k, v := range delta {
		switch {
		case strings.HasPrefix(k, session.KeyPrefixApp):
			app[strings.TrimPrefix(k, session.KeyPrefixApp)] = v
		case strings.HasPrefix(k, session.KeyPrefixUser):
			user[strings.TrimPrefix(k, session.KeyPrefixUser)] = v
		case !strings.HasPrefix(k, session.KeyPrefixTemp):
			sess[k] = v
		}
	}
	return app, user, sess
}

func mergeStates(app, user, sess map[string]any) map[string]any {
	out := make(map[string]any, len(app)+len(user)+len(sess))
	for k, v := range sess {
		out[k] = v
	}
	for k, v := range app {
		out[session.KeyPrefixApp+k] = v
	}
	for k, v := range user {
		out[session.KeyPrefixUser+k] = v
	}
	return out
}

// --- collections ---

func (s *firestoreSessionService) sessionRef(app, user, sid string) *firestore.DocumentRef {
	return s.client.Collection(s.sessions).Doc(encodeKey(app, user, sid))
}

func (s *firestoreSessionService) appStateRef(app string) *firestore.DocumentRef {
	return s.client.Collection(s.appSt).Doc(encodeKey(app))
}

func (s *firestoreSessionService) userStateRef(app, user string) *firestore.DocumentRef {
	return s.client.Collection(s.userSt).Doc(encodeKey(app, user))
}

func loadState(ctx context.Context, ref *firestore.DocumentRef) (map[string]any, error) {
	snap, err := ref.Get(ctx)
	if status.Code(err) == codes.NotFound {
		return map[string]any{}, nil
	}
	if err != nil {
		return nil, err
	}
	var d stateDoc
	if err := snap.DataTo(&d); err != nil {
		return nil, err
	}
	if d.State == nil {
		d.State = map[string]any{}
	}
	return d.State, nil
}

// loadStateTx reads a scoped state doc inside a transaction (so the read participates in the
// transaction's snapshot and conflict detection). A missing doc yields an empty map.
func loadStateTx(tx *firestore.Transaction, ref *firestore.DocumentRef) (map[string]any, error) {
	snap, err := tx.Get(ref)
	if status.Code(err) == codes.NotFound {
		return map[string]any{}, nil
	}
	if err != nil {
		return nil, err
	}
	var d stateDoc
	if err := snap.DataTo(&d); err != nil {
		return nil, err
	}
	if d.State == nil {
		d.State = map[string]any{}
	}
	return d.State, nil
}

// --- Service ---

func (s *firestoreSessionService) Create(ctx context.Context, req *session.CreateRequest) (*session.CreateResponse, error) {
	if req.AppName == "" || req.UserID == "" {
		return nil, fmt.Errorf("app_name and user_id are required, got app_name: %q, user_id: %q", req.AppName, req.UserID)
	}
	sid := req.SessionID
	if sid == "" {
		sid = uuid.NewString()
	}
	appDelta, userDelta, sessDelta := extractStateDeltas(req.State)
	now := time.Now()
	ref := s.sessionRef(req.AppName, req.UserID, sid)
	appRef := s.appStateRef(req.AppName)
	userRef := s.userStateRef(req.AppName, req.UserID)

	// One transaction creates the session and merges app/user state together, so a state
	// write failure can no longer leave a session persisted without its state (or vice
	// versa). All reads precede all writes, as Firestore transactions require.
	var appState, userState map[string]any
	err := s.client.RunTransaction(ctx, func(_ context.Context, tx *firestore.Transaction) error {
		snap, err := tx.Get(ref)
		if err == nil && snap.Exists() {
			return fmt.Errorf("session %s already exists", sid)
		}
		if err != nil && status.Code(err) != codes.NotFound {
			return err
		}
		if appState, err = loadStateTx(tx, appRef); err != nil {
			return err
		}
		if userState, err = loadStateTx(tx, userRef); err != nil {
			return err
		}
		for k, v := range appDelta {
			appState[k] = v
		}
		for k, v := range userDelta {
			userState[k] = v
		}
		if err := tx.Create(ref, sessionDoc{
			AppName: req.AppName, UserID: req.UserID, SessionID: sid,
			State: sessDelta, UpdatedAt: now,
		}); err != nil {
			return err
		}
		if len(appDelta) > 0 {
			if err := tx.Set(appRef, stateDoc{State: appState}); err != nil {
				return err
			}
		}
		if len(userDelta) > 0 {
			if err := tx.Set(userRef, stateDoc{State: userState}); err != nil {
				return err
			}
		}
		return nil
	})
	if err != nil {
		return nil, err
	}
	return &session.CreateResponse{Session: &fsSession{
		appName: req.AppName, userID: req.UserID, sessionID: sid,
		state: mergeStates(appState, userState, sessDelta), updated: now,
	}}, nil
}

func (s *firestoreSessionService) Get(ctx context.Context, req *session.GetRequest) (*session.GetResponse, error) {
	if req.AppName == "" || req.UserID == "" || req.SessionID == "" {
		return nil, fmt.Errorf("app_name, user_id, session_id are required, got app_name: %q, user_id: %q, session_id: %q", req.AppName, req.UserID, req.SessionID)
	}
	snap, err := s.sessionRef(req.AppName, req.UserID, req.SessionID).Get(ctx)
	if status.Code(err) == codes.NotFound {
		return nil, fmt.Errorf("session %s not found", req.SessionID)
	}
	if err != nil {
		return nil, err
	}
	sess, err := s.hydrate(ctx, snap)
	if err != nil {
		return nil, err
	}
	events, err := loadEvents(ctx, snap.Ref)
	if err != nil {
		return nil, err
	}
	sess.events = filterEvents(events, req.NumRecentEvents, req.After)
	return &session.GetResponse{Session: sess}, nil
}

// loadEvents reads a session's events sub-collection in sequence order.
func loadEvents(ctx context.Context, sessionRef *firestore.DocumentRef) ([]*session.Event, error) {
	docs, err := sessionRef.Collection("events").OrderBy("seq", firestore.Asc).Documents(ctx).GetAll()
	if err != nil {
		return nil, err
	}
	events := make([]*session.Event, 0, len(docs))
	for _, snap := range docs {
		var ed eventDoc
		if err := snap.DataTo(&ed); err != nil {
			return nil, err
		}
		var ev session.Event
		if err := json.Unmarshal([]byte(ed.Blob), &ev); err != nil {
			return nil, fmt.Errorf("unmarshal event: %w", err)
		}
		events = append(events, &ev)
	}
	return events, nil
}

func (s *firestoreSessionService) List(ctx context.Context, req *session.ListRequest) (*session.ListResponse, error) {
	if req.AppName == "" {
		return nil, fmt.Errorf("app_name is required, got app_name: %q", req.AppName)
	}
	q := s.client.Collection(s.sessions).Where("app_name", "==", req.AppName)
	if req.UserID != "" {
		q = q.Where("user_id", "==", req.UserID)
	}
	docs, err := q.Documents(ctx).GetAll()
	if err != nil {
		return nil, err
	}
	out := make([]session.Session, 0, len(docs))
	for _, snap := range docs {
		sess, err := s.hydrate(ctx, snap)
		if err != nil {
			return nil, err
		}
		sess.events = nil // List returns sessions without their event history
		out = append(out, sess)
	}
	return &session.ListResponse{Sessions: out}, nil
}

func (s *firestoreSessionService) Delete(ctx context.Context, req *session.DeleteRequest) error {
	if req.AppName == "" || req.UserID == "" || req.SessionID == "" {
		return fmt.Errorf("app_name, user_id, session_id are required, got app_name: %q, user_id: %q, session_id: %q", req.AppName, req.UserID, req.SessionID)
	}
	ref := s.sessionRef(req.AppName, req.UserID, req.SessionID)
	// Firestore does not cascade: delete the events sub-collection before the session doc.
	evs, err := ref.Collection("events").Documents(ctx).GetAll()
	if err != nil {
		return err
	}
	for _, ev := range evs {
		if _, err := ev.Ref.Delete(ctx); err != nil {
			return err
		}
	}
	_, err = ref.Delete(ctx)
	return err
}

func (s *firestoreSessionService) AppendEvent(ctx context.Context, curSession session.Session, event *session.Event) error {
	if curSession == nil {
		return fmt.Errorf("session is nil")
	}
	if event == nil {
		return fmt.Errorf("event is nil")
	}
	if event.Partial {
		return nil // partial events are not persisted
	}
	sess, ok := curSession.(*fsSession)
	if !ok {
		return fmt.Errorf("unexpected session type %T", curSession)
	}

	appDelta, userDelta, sessDelta := extractStateDeltas(event.Actions.StateDelta)
	stored := trimTempState(event)
	blob, err := json.Marshal(stored)
	if err != nil {
		return fmt.Errorf("marshal event: %w", err)
	}
	ref := s.sessionRef(sess.appName, sess.userID, sess.sessionID)
	appRef := s.appStateRef(sess.appName)
	userRef := s.userStateRef(sess.appName, sess.userID)
	// Merge app/user state in the same transaction as the session + event write, so the
	// scoped state can no longer advance without the event that produced it (or vice versa).
	// All reads precede all writes, as Firestore transactions require.
	err = s.client.RunTransaction(ctx, func(_ context.Context, tx *firestore.Transaction) error {
		snap, err := tx.Get(ref)
		if status.Code(err) == codes.NotFound {
			return fmt.Errorf("session not found, cannot apply event")
		}
		if err != nil {
			return err
		}
		appState, err := loadStateTx(tx, appRef)
		if err != nil {
			return err
		}
		userState, err := loadStateTx(tx, userRef)
		if err != nil {
			return err
		}
		var d sessionDoc
		if err := snap.DataTo(&d); err != nil {
			return err
		}
		if d.State == nil {
			d.State = map[string]any{}
		}
		for k, v := range sessDelta {
			d.State[k] = v
		}
		seq := d.NextSeq
		d.NextSeq++
		d.UpdatedAt = event.Timestamp
		if err := tx.Set(ref, d); err != nil {
			return err
		}
		if len(appDelta) > 0 {
			for k, v := range appDelta {
				appState[k] = v
			}
			if err := tx.Set(appRef, stateDoc{State: appState}); err != nil {
				return err
			}
		}
		if len(userDelta) > 0 {
			for k, v := range userDelta {
				userState[k] = v
			}
			if err := tx.Set(userRef, stateDoc{State: userState}); err != nil {
				return err
			}
		}
		evRef := ref.Collection("events").Doc(fmt.Sprintf("%020d", seq))
		return tx.Set(evRef, eventDoc{Seq: seq, Timestamp: event.Timestamp, Blob: string(blob)})
	})
	if err != nil {
		return err
	}

	// Reflect the append on the caller's in-memory session, mirroring the in-memory service:
	// the full (temp-stripped) delta, so app:/user: prefixed keys are visible too.
	for k, v := range stored.Actions.StateDelta {
		sess.state[k] = v
	}
	sess.events = append(sess.events, stored)
	sess.updated = event.Timestamp
	return nil
}

// hydrate loads a stored session doc plus its app/user state into an *fsSession with the
// merged, scope-prefixed state the caller expects.
func (s *firestoreSessionService) hydrate(ctx context.Context, snap *firestore.DocumentSnapshot) (*fsSession, error) {
	var d sessionDoc
	if err := snap.DataTo(&d); err != nil {
		return nil, err
	}
	appState, err := loadState(ctx, s.appStateRef(d.AppName))
	if err != nil {
		return nil, err
	}
	userState, err := loadState(ctx, s.userStateRef(d.AppName, d.UserID))
	if err != nil {
		return nil, err
	}
	if d.State == nil {
		d.State = map[string]any{}
	}
	// Events are loaded separately (Get) or omitted (List); hydrate only resolves state.
	return &fsSession{
		appName: d.AppName, userID: d.UserID, sessionID: d.SessionID,
		state: mergeStates(appState, userState, d.State), updated: d.UpdatedAt,
	}, nil
}

// trimTempState returns a copy of the event with temp: state-delta keys removed (they are
// never persisted).
func trimTempState(event *session.Event) *session.Event {
	if len(event.Actions.StateDelta) == 0 {
		return event
	}
	filtered := make(map[string]any, len(event.Actions.StateDelta))
	for k, v := range event.Actions.StateDelta {
		if !strings.HasPrefix(k, session.KeyPrefixTemp) {
			filtered[k] = v
		}
	}
	cp := *event
	cp.Actions.StateDelta = filtered
	return &cp
}

// filterEvents applies the Get request's NumRecentEvents / After filters. Events are stored
// (and loaded) in sequence order, which is not guaranteed to be timestamp order — the caller
// sets each event's Timestamp — so the After filter scans linearly rather than binary-searching
// on a monotonicity it cannot assume.
func filterEvents(events []*session.Event, numRecent int, after time.Time) []*session.Event {
	if numRecent > 0 {
		if start := len(events) - numRecent; start > 0 {
			events = events[start:]
		}
	}
	if !after.IsZero() && len(events) > 0 {
		keep := len(events)
		for i, ev := range events {
			if !ev.Timestamp.Before(after) {
				keep = i
				break
			}
		}
		events = events[keep:]
	}
	return events
}

// --- session.Session / State / Events implementations ---

type fsSession struct {
	appName, userID, sessionID string
	state                      map[string]any
	events                     []*session.Event
	updated                    time.Time
}

func (s *fsSession) ID() string                { return s.sessionID }
func (s *fsSession) AppName() string           { return s.appName }
func (s *fsSession) UserID() string            { return s.userID }
func (s *fsSession) LastUpdateTime() time.Time { return s.updated }
func (s *fsSession) State() session.State      { return &fsState{m: s.state} }
func (s *fsSession) Events() session.Events    { return fsEvents(s.events) }

type fsState struct{ m map[string]any }

func (s *fsState) Get(key string) (any, error) {
	if v, ok := s.m[key]; ok {
		return v, nil
	}
	return nil, session.ErrStateKeyNotExist
}

func (s *fsState) Set(key string, value any) error {
	s.m[key] = value
	return nil
}

func (s *fsState) All() iter.Seq2[string, any] {
	return func(yield func(string, any) bool) {
		for k, v := range s.m {
			if !yield(k, v) {
				return
			}
		}
	}
}

type fsEvents []*session.Event

func (e fsEvents) All() iter.Seq[*session.Event] {
	return func(yield func(*session.Event) bool) {
		for _, ev := range e {
			if !yield(ev) {
				return
			}
		}
	}
}

func (e fsEvents) Len() int { return len(e) }

func (e fsEvents) At(i int) *session.Event {
	if i >= 0 && i < len(e) {
		return e[i]
	}
	return nil
}

File: javascript/src/agent/setup/session_firestore.ts
/**
 * A custom ADK session service backed by Cloud Firestore — the cloud-durable session store
 * (adk-js ships only in-memory/database/vertex services). It mirrors the in-memory
 * service's semantics: app:/user:/temp: state scopes, partial-event skipping, and event
 * filtering on read.
 *
 * Events are stored in an `events` sub-collection (one JSON-blob doc per event, ordered by
 * sequence) rather than an array field, so a long-lived session cannot blow Firestore's
 * 1 MiB per-document limit. App and user state live in their own collections; the session
 * doc holds only session-scoped state. Create and appendEvent run in a transaction so the
 * session, its scoped state, and its events advance together.
 *
 * The `@google-cloud/firestore` client is loaded lazily through a real Node require (the
 * type side comes from an erased `import type`) so only this backend pulls it in. This file
 * is excluded from the default coverage gate and exercised only under the firestore emulator.
 */
import { randomUUID } from 'node:crypto';
import { createRequire } from 'node:module';
import type { DocumentReference, Firestore, Transaction } from '@google-cloud/firestore';

import {
  BaseSessionService,
  type CreateSessionRequest,
  type DeleteSessionRequest,
  type Event,
  type GetSessionRequest,
  type ListSessionsRequest,
  type ListSessionsResponse,
  type Session,
  createSession,
  mergeStates,
  State,
  trimTempDeltaState,
} from '@google/adk';

const nodeRequire = createRequire(import.meta.url);

type StateMap = Record<string, unknown>;

interface SessionDoc {
  app_name: string;
  user_id: string;
  session_id: string;
  state: StateMap;
  next_seq: number;
  updated_at: number; // epoch ms of the last event
}

interface EventDoc {
  seq: number;
  timestamp: number; // epoch ms (for the afterTimestamp filter)
  blob: string; // JSON-encoded Event
}

const KEY_DELIM = '\x1f';

/** A Firestore-safe document id: base64url of the delimiter-joined parts. */
function encodeKey(...parts: string[]): string {
  return Buffer.from(parts.join(KEY_DELIM)).toString('base64url');
}

/**
 * Split a state delta into its app / user / session scopes, stripping the scope prefix.
 * temp: keys are dropped (never persisted), mirroring the in-memory service.
 */
function extractStateDeltas(delta: StateMap): { app: StateMap; user: StateMap; sess: StateMap } {
  const app: StateMap = {};
  const user: StateMap = {};
  const sess: StateMap = {};
  for (const [k, v] of Object.entries(delta)) {
    if (k.startsWith(State.APP_PREFIX)) {
      app[k.slice(State.APP_PREFIX.length)] = v;
    } else if (k.startsWith(State.USER_PREFIX)) {
      user[k.slice(State.USER_PREFIX.length)] = v;
    } else if (!k.startsWith(State.TEMP_PREFIX)) {
      sess[k] = v;
    }
  }
  return { app, user, sess };
}

/** Apply the Get request's numRecentEvents / afterTimestamp filters (events are seq-ordered). */
function filterEvents(events: Event[], numRecent?: number, after?: number): Event[] {
  let out = events;
  if (numRecent && numRecent > 0 && out.length > numRecent) {
    out = out.slice(out.length - numRecent);
  }
  if (after && after > 0) {
    out = out.filter((e) => e.timestamp >= after);
  }
  return out;
}

/** A Firestore-backed ADK session service. */
export class FirestoreSessionService extends BaseSessionService {
  private readonly db: Firestore;
  private readonly sessionsColl: string;
  private readonly appStateColl: string;
  private readonly userStateColl: string;

  constructor(project: string, prefix: string) {
    super();
    const { Firestore: FirestoreCtor } = nodeRequire(
      '@google-cloud/firestore',
    ) as typeof import('@google-cloud/firestore');
    this.db = project ? new FirestoreCtor({ projectId: project }) : new FirestoreCtor();
    this.sessionsColl = `${prefix}_sessions`;
    this.appStateColl = `${prefix}_app_state`;
    this.userStateColl = `${prefix}_user_state`;
  }

  /** Release the underlying Firestore client. */
  async close(): Promise<void> {
    await this.db.terminate();
  }

  /** Doc reference for a single session, keyed by (app, user, session id). */
  private sessionRef(app: string, user: string, sid: string): DocumentReference {
    return this.db.collection(this.sessionsColl).doc(encodeKey(app, user, sid));
  }

  /** Doc reference for an app's shared state. */
  private appStateRef(app: string): DocumentReference {
    return this.db.collection(this.appStateColl).doc(encodeKey(app));
  }

  /** Doc reference for a user's per-app state. */
  private userStateRef(app: string, user: string): DocumentReference {
    return this.db.collection(this.userStateColl).doc(encodeKey(app, user));
  }

  /** Read a state map from a doc inside a transaction, defaulting to empty when the doc is absent. */
  private async loadStateInTx(tx: Transaction, ref: DocumentReference): Promise<StateMap> {
    const snap = await tx.get(ref);
    if (!snap.exists) {
      return {};
    }
    return ((snap.data() as { state?: StateMap }).state ?? {}) as StateMap;
  }

  /** Read a state map from a doc (non-transactional), defaulting to empty when the doc is absent. */
  private async loadState(ref: DocumentReference): Promise<StateMap> {
    const snap = await ref.get();
    if (!snap.exists) {
      return {};
    }
    return ((snap.data() as { state?: StateMap }).state ?? {}) as StateMap;
  }

  /**
   * Create a session and merge any app/user state deltas in one transaction, so a session is never
   * persisted without its state (or vice versa). Generates a session id when the request omits one.
   */
  async createSession(req: CreateSessionRequest): Promise<Session> {
    if (!req.appName || !req.userId) {
      throw new Error(
        `appName and userId are required, got appName=${JSON.stringify(req.appName)} userId=${JSON.stringify(req.userId)}`,
      );
    }
    const sid = req.sessionId && req.sessionId !== '' ? req.sessionId : randomUUID();
    const { app: appDelta, user: userDelta, sess: sessDelta } = extractStateDeltas(req.state ?? {});
    const now = Date.now();
    const ref = this.sessionRef(req.appName, req.userId, sid);
    const appRef = this.appStateRef(req.appName);
    const userRef = this.userStateRef(req.appName, req.userId);

    // One transaction creates the session and merges app/user state, so a state write can no
    // longer leave a session persisted without its state (or vice versa). All reads precede
    // all writes, as Firestore transactions require.
    const { appState, userState } = await this.db.runTransaction(async (tx) => {
      const existing = await tx.get(ref);
      if (existing.exists) {
        throw new Error(`session ${sid} already exists`);
      }
      const appState = await this.loadStateInTx(tx, appRef);
      const userState = await this.loadStateInTx(tx, userRef);
      Object.assign(appState, appDelta);
      Object.assign(userState, userDelta);

      const doc: SessionDoc = {
        app_name: req.appName,
        user_id: req.userId,
        session_id: sid,
        state: sessDelta,
        next_seq: 0,
        updated_at: now,
      };
      tx.create(ref, doc);
      if (Object.keys(appDelta).length > 0) {
        tx.set(appRef, { state: appState });
      }
      if (Object.keys(userDelta).length > 0) {
        tx.set(userRef, { state: userState });
      }
      return { appState, userState };
    });

    return createSession({
      id: sid,
      appName: req.appName,
      userId: req.userId,
      state: mergeStates(appState, userState, sessDelta),
      events: [],
      lastUpdateTime: now,
    });
  }

  /** Load a session with its (optionally filtered) event history, merged over app/user/session state. */
  async getSession(req: GetSessionRequest): Promise<Session | undefined> {
    const snap = await this.sessionRef(req.appName, req.userId, req.sessionId).get();
    if (!snap.exists) {
      return undefined;
    }
    const doc = snap.data() as SessionDoc;
    const events = await this.loadEvents(snap.ref);
    const filtered = filterEvents(events, req.config?.numRecentEvents, req.config?.afterTimestamp);
    const appState = await this.loadState(this.appStateRef(req.appName));
    const userState = await this.loadState(this.userStateRef(req.appName, req.userId));
    return createSession({
      id: req.sessionId,
      appName: req.appName,
      userId: req.userId,
      state: mergeStates(appState, userState, doc.state ?? {}),
      events: filtered,
      lastUpdateTime: doc.updated_at,
    });
  }

  /** Read a session's events from its `events` sub-collection, ordered by sequence number. */
  private async loadEvents(sessionRef: DocumentReference): Promise<Event[]> {
    const docs = await sessionRef.collection('events').orderBy('seq', 'asc').get();
    return docs.docs.map((d) => JSON.parse((d.data() as EventDoc).blob) as Event);
  }

  /** List an app's sessions (optionally scoped to a user), each without its event history. */
  async listSessions(req: ListSessionsRequest): Promise<ListSessionsResponse> {
    if (!req.appName) {
      throw new Error(`appName is required, got ${JSON.stringify(req.appName)}`);
    }
    let query = this.db.collection(this.sessionsColl).where('app_name', '==', req.appName);
    if (req.userId) {
      query = query.where('user_id', '==', req.userId);
    }
    const snap = await query.get();
    const sessions: Session[] = [];
    for (const d of snap.docs) {
      const doc = d.data() as SessionDoc;
      const appState = await this.loadState(this.appStateRef(doc.app_name));
      const userState = await this.loadState(this.userStateRef(doc.app_name, doc.user_id));
      sessions.push(
        createSession({
          id: doc.session_id,
          appName: doc.app_name,
          userId: doc.user_id,
          state: mergeStates(appState, userState, doc.state ?? {}),
          events: [], // List returns sessions without their event history
          lastUpdateTime: doc.updated_at,
        }),
      );
    }
    const totalItems = sessions.length;
    return { sessions, page: 1, limit: totalItems, totalItems, totalPages: totalItems === 0 ? 0 : 1 };
  }

  /** Delete a session and its events sub-collection (Firestore does not cascade). */
  async deleteSession(req: DeleteSessionRequest): Promise<void> {
    const ref = this.sessionRef(req.appName, req.userId, req.sessionId);
    // Firestore does not cascade: delete the events sub-collection before the session doc.
    const events = await ref.collection('events').get();
    for (const ev of events.docs) {
      await ev.ref.delete();
    }
    await ref.delete();
  }

  override async appendEvent({ session, event }: { session: Session; event: Event }): Promise<Event> {
    if (event.partial) {
      return event; // partial events are not persisted
    }
    const stored = trimTempDeltaState(event);
    const { app: appDelta, user: userDelta, sess: sessDelta } = extractStateDeltas(
      stored.actions?.stateDelta ?? {},
    );
    const blob = JSON.stringify(stored);
    const ref = this.sessionRef(session.appName, session.userId, session.id);
    const appRef = this.appStateRef(session.appName);
    const userRef = this.userStateRef(session.appName, session.userId);

    await this.db.runTransaction(async (tx) => {
      const snap = await tx.get(ref);
      if (!snap.exists) {
        throw new Error('session not found, cannot apply event');
      }
      const appState = await this.loadStateInTx(tx, appRef);
      const userState = await this.loadStateInTx(tx, userRef);
      const doc = snap.data() as SessionDoc;
      const sessionState = doc.state ?? {};
      Object.assign(sessionState, sessDelta);
      const seq = doc.next_seq;

      tx.set(ref, {
        ...doc,
        state: sessionState,
        next_seq: seq + 1,
        updated_at: event.timestamp,
      } satisfies SessionDoc);
      if (Object.keys(appDelta).length > 0) {
        Object.assign(appState, appDelta);
        tx.set(appRef, { state: appState });
      }
      if (Object.keys(userDelta).length > 0) {
        Object.assign(userState, userDelta);
        tx.set(userRef, { state: userState });
      }
      const evRef = ref.collection('events').doc(String(seq).padStart(20, '0'));
      tx.set(evRef, { seq, timestamp: event.timestamp, blob } satisfies EventDoc);
    });

    // Reflect the append on the caller's in-memory session, mirroring the in-memory service:
    // the full (temp-stripped) delta, so app:/user:-prefixed keys are visible too.
    for (const [k, v] of Object.entries(stored.actions?.stateDelta ?? {})) {
      session.state[k] = v;
    }
    session.events.push(stored);
    session.lastUpdateTime = event.timestamp;
    return stored;
  }
}
